platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/graph/builder/NodeState.java

```diff
@@ -324,37 +326,46 @@ void cleanupConstraints() {
     }
 
     private boolean excludesSameDependenciesAsPreviousTraversal(ExcludeSpec newResolutionFilter) {
-        List<DependencyState> oldStates = cachedFilteredDependencyStates;
-        if (previousTraversalExclusions == null || oldStates == null) {
+        if (previousTraversalExclusions == null) {
+            // There was no previous traversal. This traversal can't be the same.
             return false;
         }
         if (previousTraversalExclusions.equals(newResolutionFilter)) {
+            // The excludes did not change. The dependencies are the same.
             return true;
         }
+        if (cachedFilteredDependencyStates == null) {
+            // We don't know which dependencies were excluded in the previous traversal.
+            // We are not sure if this traversal is the same.
+            return false;
+        }
         if (doesNotHaveDependencies && !dependenciesMayChange) {
-            // whatever the exclude filter, there are no dependencies
+            // We have no dependencies, so the resolution filter does not matter.
             return true;
         }
-        cachedFilteredDependencyStates = null;
+
         // here, we need to check that applying the new resolution filter
         // we would actually exclude exactly the same dependencies as in
         // the previous visit. It is important that this is NOT a heuristic
         // (it used to be) because if the filters are _equivalent_, we would
         // revisit all dependencies and possibly change the classpath order!
-        boolean sameDependencies = dependencies(newResolutionFilter).equals(oldStates);
+        List<DependencyState> oldDependencies = cachedFilteredDependencyStates;
+        this.cachedFilteredDependencyStates = null; // Invalidate the cache so `dependencies` recomputes the value.
+        boolean sameDependencies = dependencies(newResolutionFilter).equals(oldDependencies);
+
         if (sameDependencies) {
-            // While there will be no change to this node, there might be changes to the nodes it brings as the exclude change could concern them
+            // The excludes did change. Even if our direct filtered dependencies are the same,
+            // their transitive dependencies may be affected by the new excludes. So, we
+            // always update the outgoing edges to reflect the new resolution filter.
+
+            // If we return true here, we will short-circuit normal dependency traversal,
+            // which usually takes care of updating the transitive excludes. Do that here
+            // instead.
```

Comment:
This all needs some updated wording now that we are no longer unconditionally doing this